File: interviewpython/interviewpython/longestnonrepsubstr.py
class longestnonrepsubstr(object):
    def findLongestNonrepSubstr(self, s):
        if (len(s) == 0):
            return 0
        map = [0]*255;
        left = 0
        ret = 0
        
        for i in range(0, len(s)):
            idx = ord(s[i]);
            if (map[idx] == 0 or map[idx] < left):
                ret = max(ret, i - left + 1)
            else:
                left = map[idx]
            map[idx] = i + 1
        
        return ret

if __name__=="__main__":
    print("PY:longestnonrepsubstr")
    solu = longestnonrepsubstr()
    s = "abbca"
    result = solu.findLongestNonrepSubstr(s)
    print(s)
    print(result)

File: interviewcpp/dp/longestnonrepsubstr/longestnonrepsubstr.cpp
using namespace std;

#include <iostream>

class longestnonrepsubstr{
public:
	int lengthOfLongestSubstring(string s) {
		int res = 0;
		int m[256] = {0};
		int left = 0;

		for (int i = 0; i < s.size(); i++) {
			if (m[s[i]] == 0 || m[s[i]] < left) {
				res = max(res, i-left+1);
			} else {
				left = m[s[i]];
			}
			m[s[i]] = i+1;  // make it different from 0
		}
		return res;
	}


};

int main(){
	cout << "CPP:lengthOfLongestSubstring" << endl;
	string A = "abbca";
	cout << A << endl;

	longestnonrepsubstr test;
	int result = test.lengthOfLongestSubstring(A);
	cout << result << endl;

	return 0;
}
